### PR TITLE
[Kernel] Use flash-attn for decoding

### PR DESCRIPTION
**Pull request by skrider**
Vendors flash-attention from https://github.com/Dao-AILab/flash-attention/pull/824, prunes out the backward pass operator for faster compile times, adds reshape and cache kernel for flash attention kv cache layout, adds logic for selecting kv cache manager / attention backend based on temporary environment variable VLLM_TEMP_USE_FLASH_DECODE. Tested for single GPU on opt-125m, llama-7b

### CODE REVIEW

**Comment by simon-mo** [comment]
We will try to get this in best effort by tomorrow, if not, this will be slated for next release

**Comment by WoosukKwon** [comment]
@skrider Thanks for the great work! Can I directly fix this PR for faster integration?

File: vllm/attention/ops/flash_attn.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        block_tables: torch.Tensor,
+        subquery_start_loc: torch.Tensor,
+        prompt_lens_tensor: torch.Tensor,
+        context_lens: torch.Tensor,
+        max_subquery_len: int,
+        alibi_slopes: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        raise NotImplementedError
```
Comment (WoosukKwon) [inline vllm/attention/ops/flash_attn.py:127]: @skrider Could you elaborate on why it is tricky to implement prefix-enabled attention?

Comment (skrider) [reply vllm/attention/ops/flash_attn.py:127]: The flash API does not support attention to dense KV and paged KV in the same launch. We can either cache the dense KV before invoking forward_prefix or compute attention separately and merge with the online softmax trick. Not sure which would be best.

Comment (rkooo567) [reply vllm/attention/ops/flash_attn.py:127]: @skrider isn't this possible with `flash_attn_varlen_func`? It seems like it supports paged kv cache (https://github.com/Dao-AILab/flash-attention/pull/831/files). 

IIUC, you can just pass k cache and v cache to k and v for `flash_attn_varlen_func` IIUC?

**Review by rkooo567 — COMMENTED** [review]
QQ: Do we really need a new backend implementation? Why don't we just adding env var to existing flash attn impl's decoding path?

File: vllm/attention/ops/flash_attn.py (hunk 1 above)
Comment (skrider) [reply vllm/attention/ops/flash_attn.py:127]: Yes this is correct, however the kernel implements this by first copying k,v into the KV cache in gmem, and then reading the KV cache and computing attention, so we would need to ensure that there is enough space in the KV cache to append. And at that point why not just copy them in beforehand as is done with forward_decode, keeping the attention operator stateless as is implied.

**Comment by skrider** [comment]
@rkooo567 I think that because the KV cache layout is different it makes sense to have a different backend.

**Comment by WoosukKwon** [comment]
@skrider I just edited this PR: 1) I removed dependency on your FlashAttention repo (Let's add it in the next PR), and 2) I moved this to `FlashAttentionBackend` as @rkooo567 suggested, since the backend actually existed for this integration. The current implementation was actually a placeholder. Could you please take a look?

File: vllm/attention/ops/flash_attn.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/attention/ops/flash_attn.py]: > We can either cache the dense KV before invoking forward_prefix

@skrider This is what our current Triton kernel is doing. Can't we do the same thing with flash-attn? Please take a look at the change I made in this PR.

**Comment by zhaoyang-star** [comment]
Unittest is needed.

**Comment by WoosukKwon** [comment]
Status update:

1. I created `vllm-flash-attn` package in PyPI based on @skrider's small-page version of `flash-attn` (Repo: https://github.com/vllm-project/flash-attention). The package was built with PyTorch 2.1.2 and CUDA 12.1. I cut down the binary size to 98 MB by removing the kernels for backward pass.
2. When tested on Mistral, this PR gives ~10% throughput increase on ShareGPT, thanks to the GQA optimization.
3. Further tests are required, especially for prefix-enabled attention.
4. Also, I'm not sure if FlashDecoding is used automatically or we should manually turn it on.

File: vllm/attention/ops/flash_attn.py (hunk 1 above)
Comment (skrider) [reply vllm/attention/ops/flash_attn.py]: My mistake - we should be good then

File: vllm/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -207,32 +234,32 @@ def forward(
...
-                attn_metadata.context_lens,
-                attn_metadata.max_context_len,
-                attn_metadata.kv_cache_dtype,
-                self.num_kv_heads,
-                self.scale,
-                self.alibi_slopes,
+                block_table=attn_metadata.block_tables,
+                cache_seqlens=attn_metadata.context_lens,
+                softmax_scale=self.scale,
+                causal=True,
```
Comment (skrider) [inline vllm/attention/backends/flash_attn.py:261]: Nit: redundant when seqlen_q = 1

**Comment by skrider** [comment]
LGTM :+1:

File: vllm/attention/backends/flash_attn.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py:261]: Good point. Right. I just added it for clarity.

File: vllm/attention/backends/flash_attn.py [hunk 3]
```diff
@@ -207,32 +234,32 @@ def forward(
...
+                alibi_slopes=self.alibi_slopes,
```
Comment (rkooo567) [inline vllm/attention/backends/flash_attn.py:262]: should we allow to specify num_splits using env var? I think it is automatically chosen, but wonder if we should expose this

**Review by rkooo567 — APPROVED** [review]
minor comments regarding cache update unit test!

**Comment by simon-mo** [comment]
what's the status of this PR?

**Comment by WoosukKwon** [comment]
@simon-mo It's not ready because it doesn't produce the correct results when using attention with prefix. Also, this PR needs more tests. It'll be hard to include this in v0.4.0 unfortunately.

**Comment by WoosukKwon** [comment]
Related PRs: #2744 and #3010

File: csrc/cache_kernels.cu [hunk 4]
```diff
@@ -209,6 +209,42 @@ __global__ void reshape_and_cache_kernel(
...
+  for (int i = threadIdx.x; i < n; i += blockDim.x) {
+    const int64_t src_key_idx = token_idx * key_stride + i;
+    const int64_t src_value_idx = token_idx * value_stride + i;
+    const int head_idx = i / head_size;
+    const int head_offset = i % head_size;
+    const int64_t tgt_value_idx = block_idx * block_stride
+                              + block_offset * num_heads * head_size
+                              + head_idx * head_size
+                              + head_offset;
+    k_cache[tgt_value_idx] = key[src_key_idx];
```
[the remark this answers is not in the document]
Comment (skrider) [reply csrc/cache_kernels.cu:243]: Flash Attention currently does not support e5m2, so I didn't worry about it. Good point though, will add a check.

**Comment by rkooo567** [comment]
QQ: is this PR still active?

**Comment by rkooo567** [comment]
Btw, it also seems like vllm flash attn is built with glib 2.32, and is not compatible with ubuntu 20.04 (which has glib 2.31 by default).

File: vllm/attention/backends/flash_attn.py [hunk 5]
```diff
@@ -234,37 +255,35 @@ def forward(
...
-                    prefill_meta.prompt_lens_tensor,
-                    prefill_meta.context_lens,
-                    prefill_meta.max_subquery_len,
-                    self.alibi_slopes,
+                # FIXME(woosuk): FlashAttention does not support FP8 KV cache.
+                output[:num_prefill_tokens] = flash_attn_varlen_func(
+                    q=query,
+                    k=key_cache,
+                    v=value_cache,
+                    cu_seqlens_q=prefill_meta.seq_start_loc,
```
Comment (rkooo567) [inline vllm/attention/backends/flash_attn.py:263]: I think this needs to be 

                    prefill_meta.subquery_start_loc,
                    prefill_meta.seq_start_loc,
                    prefill_meta.max_subquery_len,
                    prefill_meta.max_prompt_len,

Comment (rkooo567) [reply vllm/attention/backends/flash_attn.py:263]: https://github.com/vllm-project/vllm/pull/4560/commits/bf3abdc9e3cf044453ac5954f84c324f55c06501

Verified locally that this makes correctness test pass

File: vllm/attention/backends/flash_attn.py [hunk 6]
```diff
@@ -143,10 +166,10 @@ def __init__(
         assert self.num_heads % self.num_kv_heads == 0
         self.num_queries_per_kv = self.num_heads // self.num_kv_heads
 
-        suppored_head_sizes = PagedAttention.get_supported_head_sizes()
+        suppored_head_sizes = [32, 64, 96, 128, 160, 192, 224, 256]
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py]: Good point. Fixed!

File: vllm/attention/backends/flash_attn.py [hunk 7]
```diff
@@ -33,27 +27,36 @@ def get_kv_cache_shape(
         num_kv_heads: int,
         head_size: int,
     ) -> Tuple[int, ...]:
-        return PagedAttention.get_kv_cache_shape(num_blocks, block_size,
-                                                 num_kv_heads, head_size)
+        if block_size != 16:
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py]: nice catch! Fixed.

File: vllm/attention/backends/flash_attn.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py:284]: As you might know, the `flash-attn` library includes heuristics to choose `num_splits`. Since this value highly depends on the input shapes, it does not make sense to set this as an env variable.

File: vllm/attention/backends/flash_attn.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py]: Thanks for bringing this up. I made the similar change along with some changes in Model Runner.

**Review by rkooo567 — APPROVED** [review]
Looks good! Btw, one issue I found is that the vllm flash attn is not working at ubuntu 20.04. Maybe we should build the wheel in 20.04 instead of 22.04? We should probably rebuild the wheel using torch 2.3 as well

Also, https://github.com/Dao-AILab/flash-attention/pull/824 -> the author is willing to merge this PR (@skrider ), and it'd be great if we can use the official wheel!

**Comment by WoosukKwon** [comment]
@rkooo567 Thanks for letting me know that the wheel doesn't work on ubuntu 20.04. Let me fix this before merge. The wheel is already built on torch 2.3 btw.

> Also, https://github.com/Dao-AILab/flash-attention/pull/824 -> the author is willing to merge this PR (@skrider ), and it'd be great if we can use the official wheel!

Unfortunately, we cannot use the official wheel because of pypi constraints. The flash-attn wheel exceeds the pypi size limit (100 MiB) and thus uses source distribution instead of build distribution. And its source distribution package does not work well on some environments. The `vllm-flash-attn` is a workaround for this problem: I pruned the binaries for backward pass, so it fits into the size limit.

**Comment by rkooo567** [comment]
I see. that makes a lot of sense! I think vendoring in the current way should be fine! 

Besides, if the wheel is available at 20.04, we can also run it in a100 or h100 and share the result. I expect to improve chunked perfill (last time I checked, prefix kernel is at least 3X faster with flash attn) / llama 70B (due to GQA) perf!

File: vllm/attention/backends/flash_attn.py [hunk 8]
```diff
@@ -234,38 +257,34 @@ def forward(
...
-                attn_metadata.kv_cache_dtype,
-                self.num_kv_heads,
-                self.scale,
-                self.alibi_slopes,
-                kv_scale,
-            )
+                block_table=decode_meta.block_tables,
+                cache_seqlens=decode_meta.seq_lens_tensor,
+                softmax_scale=self.scale,
+                causal=True,
```
Comment (rkooo567) [inline vllm/attention/backends/flash_attn.py:285]: We should probably support window_size here?

**Comment by zhisbug** [comment]
@sfc-gh-aqiao  watch this thread -- which influence chunk-prefill performance

File: vllm/attention/backends/flash_attn.py (hunk 8 above)
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn.py:289]: Good point. First of all, `sliding_window` is not considered here because we manage it through adjusting the block table. This is the same for other attention backends which use `PagedAttention`.

Second, that being said, actually it seems sliding window is not properly supported in the prefill + paged KV cache case. For a temporary workaround, I changed the selection logic to fall back to `xFormersBackend` when `sliding_window` is not None.

File: vllm/worker/model_runner.py [hunk 9]
```diff
@@ -266,20 +266,23 @@ def _prepare_prompt(
                 # Prefix is not supported with sliding_window
                 context_len = len(computed_block_nums) * self.block_size
                 prompt_tokens = prompt_tokens[context_len:]
-                prefix_block_tables.append(computed_block_nums)
+                if self.attn_backend.get_name() == "flash-attn":
+                    block_table = seq_group_metadata.block_tables[seq_id]
+                else:
+                    block_table = computed_block_nums
```
Comment (Yard1) [inline vllm/worker/model_runner.py:272]: this suggests to me we should use some abstraction here instead of if...else branching

File: tests/kernels/test_flash_attn.py [hunk 10]
```diff
@@ -45,6 +46,13 @@ def ref_paged_attn(
         attn = torch.einsum("qhd,khd->hqk", q, k)
         mask = torch.triu(torch.ones(query_len, kv_len),
                           diagonal=kv_len - query_len + 1).bool()
+        # print(mask)
+        if sliding_window is not None:
+            sliding_window_mask = torch.triu(torch.ones(query_len, kv_len),
```
Comment (Yard1) [inline tests/kernels/test_flash_attn.py:50]: reuse the torch.ones call from above?

Comment (WoosukKwon) [reply tests/kernels/test_flash_attn.py]: Fixed!

File: vllm/worker/model_runner.py (hunk 9 above)
Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: Good point. Agreed. For now, I added a note on why we need this `if`, and added TODO asking for a better abstraction.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py]: I think the simplest solution here is to fix the Triton kernel so that it aligns with the other backends' APIs. I will leave this for future PR though.

**Comment by rkooo567** [comment]
oh yay! I will run the benchmark on a100 today

**Comment by wooyeonlee0** [comment]
Great! Thanks for the PR. @skrider 

It seems that this PR has introduced the following constraint.
`_SUPPORTED_HEAD_SIZES = [32, 64, 96, 128, 160, 192, 224, 256]` [link](https://github.com/vllm-project/vllm/pull/4907/files#diff-c310ada35beeefacf4f019051ceaffeb471117d5d5b8be51610d80c7632c6bdcR12)

I've been using vllm for testing the model with head size not listed in the above list.

Then from now on, can't I run my model on vllm with the flash_attn option?
Does vllm before this PR (v0.4.2) work properly with the model with head size not listed in the list?

I look forward to hearing from you.
Thank you in advance!

**Comment by rkooo567** [comment]
I think if this head size is not supported, you cannot use the flash attn (it is the limitation of flash attn).

**Comment by rkooo567** [comment]
(to make it work, you should probably make flash attn work with unspecified head sizes)

**Comment by wooyeonlee0** [comment]
@rkooo567 Oh, I didn't know that flash attn has this limitation. Thanks for the information!
